### PR TITLE
feat: allow configuring display apps from template

### LINKED ISSUE #8033: Make helper apps optional (e.g. “VS Code Desktop”)
Version: [`v0.24.1+2c843f4`](https://github.com/coder/coder/commit/2c843f4011430ef5bbefc76460af20393a317350)

An enterprise prospect is requesting that VS Code Desktop icon be configurable. i..e, removed in certain templates.

Their feedback was that there are many development teams that never will use VS Code. e.g., Java, data science, etc. so do not want to confuse or annoy end users about using Coder.

image

Related #8032 

## Implementation notes

Written by @bpmct 

To me, the most natural way to do this is per-agent. Instead of specific flags to disable each “desktop app” (let’s assume we will have many down the road such as JetBrains, RDP, etc), I suggest we have a field on the agent that defaults to specific icons showing up.

```hcl
resource “coder_agent” “data_science” {
  # …
  helper_apps = [] # disable all
}
```

And by default, agents have this:
```hcl
resource “coder_agent” “data_science” {
  # …
  helper_apps = [“vs-code”, “port-forward”, “ssh”]
}
```

By “default,” I mean the default for the provider. Not our example templates.

One interesting side-effect of this approach is it allows template authors “opt-in” to new apps appearing on their dashboard if they hardcode the helper apps in their template OR if we don’t make it a default property for the agent. This allows us to add more functionality `vs-code-insiders` without being concerned about the impact it has on existing deployments that upgrade: “wait, my users are now seeing a new button on the dashboard I don’t want them to see”. 

Open to other ideas though!

### PR DESCRIPTION
**Pull request by sreya**
This PR implements the control plane portion of the `display_apps` field. 
- The default behavior for the field is to opt-in to all apps (except vscode_insiders).
- To opt-out of all you must specify `false` to all fields. 
- Otherwise you only get the apps that you explicitly set to `true`.

This PR implements both the backend and frontend portions of the feature. Notably on the frontend the dropdown for `VSCode Desktop` ceases to be a dropdown if only one of `VSCode Desktop` or `VSCode Insiders` is enabled.

Some screenshots:

```
displays_apps {}
``` 

Screenshot 2023-08-28 at 5 29 28 PM

```
display_apps {
  vscode = false
  vscode_insiders = true
}
```
Screenshot 2023-08-28 at 5 35 39 PM

```
display_apps {
  vscode_insiders = true
}
```
Screenshot 2023-08-28 at 5 41 18 PM

fixes #8033

### CODE REVIEW

File: provisioner/terraform/testdata/display-apps/display-apps.tf
Comment (sreya) [inline provisioner/terraform/testdata/display-apps/display-apps.tf:4]: I'm not entirely sure what to do about this, I guess I can update it to the version that the plugin _will_ be?

File: coderd/database/modelmethods.go
Comment (mtojek) [inline coderd/database/modelmethods.go:13]: Is it used anywhere?

File: coderd/provisionerdserver/provisionerdserver_test.go
Comment (mtojek) [inline coderd/provisionerdserver/provisionerdserver_test.go:1539]: I'm wondering if `DisplayApps` can refer to provisionersdk or a similar place to pull in these defaults.

File: codersdk/workspaceagents.go
Comment (mtojek) [inline codersdk/workspaceagents.go:139]: QQ: do we need to take any action on the API level based these enum values? Maybe `codersdk`/API can be just a "dumb proxy", I mean "any" app.

**Review by mtojek — COMMENTED** [review]
The first review round is done. Good job!

File: provisioner/terraform/testdata/display-apps/display-apps.tf
Comment (sreya) [reply provisioner/terraform/testdata/display-apps/display-apps.tf:4]: Yeah it's just that the feature isn't merged into the terraform provider yet [here](https://github.com/coder/terraform-provider-coder/pull/147). So I guess once I get approval here + there Ill merge the terraform one and release and update the `source` here prior to merging.

File: provisioner/terraform/testdata/resource-metadata-duplicate/resource-metadata-duplicate.tfplan.json
Comment (sreya) [reply provisioner/terraform/testdata/resource-metadata-duplicate/resource-metadata-duplicate.tfplan.json:1]: I just ran the `generate.sh` script :shrug: . I could remove all the unrelated changes but given I'm just running our dogfood image I don't think it's a big deal to go ahead and update them now. Especially if it'll keep happening to others. No strong opinion here though.

File: coderd/database/modelmethods.go
Comment (sreya) [reply coderd/database/modelmethods.go:13]: Nah a relic from an earlier implementation

File: codersdk/workspaceagents.go
Comment (sreya) [reply codersdk/workspaceagents.go:139]: No not really it's mainly for discoverability and it also helps with the generated frontend types but I also don't mind removing it, wdyt?

File: provisionersdk/agent.go
Comment (sreya) [reply provisionersdk/agent.go:42]: This is a relic, will remove

File: coderd/provisionerdserver/provisionerdserver_test.go
Comment (sreya) [reply coderd/provisionerdserver/provisionerdserver_test.go:1539]: I added this but it doesn't actually apply to this test case since it's testing to ensure all possible values are translated correctly.

File: coderd/provisionerdserver/provisionerdserver.go
Comment (sreya) [reply coderd/provisionerdserver/provisionerdserver.go:1552]: It would be nice but since they're all explicitly enumerated it makes it difficult to fail on an unknown.

File: codersdk/workspaceagents.go
Comment (mtojek) [reply codersdk/workspaceagents.go:139]: My idea is to reduce the number of places to update when you add another application, so if we don't intend to use let's just drop it. Unless we will use it in TypeScript/site, then we can leave it 👍

**Review by mtojek — APPROVED** [review]
Implementation LGTM 👍 but please take a look at e2e tests as it is failing for webTerminal? maybe it is just flake

File: provisioner/terraform/testdata/display-apps/display-apps.tf
Comment (mtojek) [reply provisioner/terraform/testdata/display-apps/display-apps.tf:4]: SGTM!

Comment (sreya) [reply provisioner/terraform/testdata/display-apps/display-apps.tf:4]: fixed

**Pull request merged**
